tsdb/querier.go

```diff
@@ -244,6 +245,41 @@ func PostingsForMatchers(ix IndexReader, ms ...*labels.Matcher) (index.Postings,
 			labelMustBeSet[m.Name] = true
 		}
 	}
+	isSubtractingMatcher := func(m *labels.Matcher) bool {
+		if !labelMustBeSet[m.Name] {
+			return true
+		}
+		return (m.Type == labels.MatchNotEqual || m.Type == labels.MatchNotRegexp) && m.Matches("")
+	}
+	someSubtractingMatchers, someIntersectingMatchers := false, false
```

Comment:
Nit: I prefer `hasSubstractingMatchers, hasIntersectingMatchers` but its fine if you don't want to change this